File: modules/xml/app/model/XmlExType.scala
package model

import model.XmlExType._

import scala.collection.mutable.ListBuffer

object XmlExType {

  val values: ListBuffer[XmlExType] = ListBuffer.empty

  def byName(n: String): Option[XmlExType] = values.find(_.name == n)

}

abstract sealed class XmlExType(val name: String, val studFileEnding: String, val refFileEnding: String, val gramFileEnding: String)
  extends model.exercise.Tag {

  values += this

  override val getButtonContent: String = name

  override def getTitle: String = {
    val forTitle = name.split("_")
    s"${forTitle(0)} gegen ${forTitle(1)}"
  }

}

case object XML_XSD extends XmlExType("XML_XSD", "xml", "xsd", "xsd")

case object XML_DTD extends XmlExType("XML_DTD", "xml", "dtd", "dtd")

case object XSD_XML extends XmlExType("XSD_XML", "xsd", "xml", "xsd")

case object DTD_XML extends XmlExType("DTD_XML", "dtd", "xml", "dtd")

File: app/model/xml/XmlExType.java
package model.xml;

import model.Enums.Selectable;
import model.ExTag;
import scala.Option;
import scala.Some;

public enum XmlExType implements Selectable<XmlExType>, ExTag {

    XML_XSD("XML", "XSD", "xml", "xsd", "xsd"),

    XML_DTD("XML", "DTD", "xml", "dtd", "dtd"),

    XSD_XML("XSD", "XML", "xsd", "xml", "xsd"),

    DTD_XML("DTD", "XML", "dtd", "xml", "dtd");

    public final String given;
    public final String toCreate;

    public final String studFileEnding;
    public final String refFileEnding;
    public final String gramFileEnding;

    XmlExType(String theGiven, String theToCreate, String theStudFileEnding, String theRefFileEnding, String theGramFileEnding) {
        given = theGiven;
        toCreate = theToCreate;
        studFileEnding = theStudFileEnding;
        refFileEnding = theRefFileEnding;
        gramFileEnding = theGramFileEnding;
    }

    @Override
    public String buttonContent() {
        return given + "_" + toCreate;
    }

    @Override
    public String title() {
        return toCreate + " gegen " + given;
    }

    public static Option<XmlExType> byString(String str) {
        try {
            return new Some<>(valueOf(str));
        } catch (Exception e) {
            // ==> None!
            return Option.apply(null);
        }
    }

}
